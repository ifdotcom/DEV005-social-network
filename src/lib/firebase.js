// Import the functions you need from the SDKs you need
import { initializeApp } from 'firebase/app';
import { getAuth, onAuthStateChanged } from 'firebase/auth';
import {
  getFirestore, doc, deleteDoc, collection, addDoc,
  getDocs, onSnapshot, orderBy, updateDoc, getDoc, arrayUnion, arrayRemove, query,
} from 'firebase/firestore';

// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// const firebaseConfig = {
//   apiKey: 'AIzaSyDr_i_8XfqSbS2ej-v0w-r6Wy2FCpmL_sw',
//   authDomain: 'social-network-b08d0.firebaseapp.com',
//   projectId: 'social-network-b08d0',
//   storageBucket: 'social-network-b08d0.appspot.com',
//   messagingSenderId: '611309449971',
//   appId: '1:611309449971:web:52cdfd8a7f8dfccac87d86',
// };
// const firebaseConfig = {
//   apiKey: 'AIzaSyBpUAmyGv1olj6CsTJ6A8rZtCvK3XgzpXA',
//   authDomain: 'myfirstproyect-bb58e.firebaseapp.com',
//   projectId: 'myfirstproyect-bb58e',
//   storageBucket: 'myfirstproyect-bb58e.appspot.com',
//   messagingSenderId: '849896345504',
//   appId: '1:849896345504:web:4482e1224f5e61429d5ec5',
//   measurementId: 'G-5HVRKB5KL9',
// };
const firebaseConfig = {
  apiKey: "AIzaSyAhTA8a9C8F5CoZGVaZqjPaVlFmIJKnmUk",
  authDomain: "frikis-del-espacio-aa6ea.firebaseapp.com",
  projectId: "frikis-del-espacio-aa6ea",
  storageBucket: "frikis-del-espacio-aa6ea.appspot.com",
  messagingSenderId: "420233466361",
  appId: "1:420233466361:web:02c7d97249f47c31fded28",
};

// Initialize Firebase
export const app = initializeApp(firebaseConfig);
export const auth = getAuth(app);
// Initialize Cloud Firestore and get a reference to the service
export const db = getFirestore(app);
// Función para guardar posts
// se crea la colección
export const savePost = (idUser, post, date, likes) => {
  addDoc(collection(db, 'posts'), {
    idUser, post, date, likes,
  });
};
export const getPosts = () => getDocs(collection(db, 'posts'));
export const onGetPosts = (callback) => onSnapshot(query(collection(db, 'posts'), orderBy('date', 'desc')), callback);
export const deletePost = (id) => deleteDoc(doc(db, 'posts', id));
export const getPost = (id) => getDoc(doc(db, 'posts', id));
export const editPost = (id, newPost) => updateDoc(doc(db, 'posts', id), newPost);
export const likePost = (id, idUser) => updateDoc(doc(db, 'posts', id), {
  likes: arrayUnion(idUser),
});
export const dislikePost = (id, idUser) => updateDoc(doc(db, 'posts', id), {
  likes: arrayRemove(idUser),
});
export const verifyUser = () => {
  onAuthStateChanged(auth, (user) => {
    console.log(user);
    if (window.location.pathname === '/dashboard' && user === null) {
      window.location.pathname = '/';
    } else if (window.location.pathname === '/' && user) {
      window.location.pathname = '/dashboard';
    } else if (window.location.pathname === '/login' && user) {
      window.location.pathname = '/dashboard';
    } else if (window.location.pathname === '/register' && user) {
      window.location.pathname = '/dashboard';
    } else if (window.location.pathname === '' && user) {
      window.location.pathname = '/dashboard';
    }
  });
};
