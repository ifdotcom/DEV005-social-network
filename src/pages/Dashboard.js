import { onAuthStateChanged } from 'firebase/auth';
import { buttonSignOut } from '../lib/SignOut.js';
import { savePostFire, gettingPosts } from '../lib/Posts.js';
import {
  deletePost,
  getPost,
  editPost,
  auth,
  likePost,
  dislikePost,
} from '../lib/firebase.js';

import vaca2 from '../img/vaca2.png';
import alien from '../img/imgAlienDashboard.png';
import ovni from '../img/ovni.gif';

const Dashboard = (navigateTo) => {
  const viewDashboard = `
    <div id="headerDashboard">
      <h1 class="tittle-dashboard">Frikis Del Espacio</h1>
      <article id= "LogOut">
        <img id="logoSignOut" src= "${vaca2}" alt="Imagen  "/>
        <button id ="button-signOut">Cerrar Sesión</button>
      </article>
    </div>
    <div id="barDashboard"></div>
    
    <div id="profile">
    <div id="circleAlien">
      <div id= "circleAlien2">
        <img id="imgAlien" src= "${alien}" alt="Imagen de Alien "/>
      </div>
    </div>
    <span id= "userName"> Gabriela79 </span>
    </div>
    <div id = "border-textArea">
      <textarea id="post-text" name="textarea" placeholder="Escribir publicación..."></textarea>
    </div>
    <div id = "msg-post">
      <span id="msg-post-text">Se ha compartido tu publicación</span>
    </div>
    <div id="container-btn">
      <button id="button-post" class="btn-post">Publicar</button>
    </div>
    <div id='containerPosts'>
      <div class="box-gradient">
        <div id="postPublic">
          <span class="loading">Cargando...</span>
          <img id="ovniL" src= "${ovni}" alt="Ovni girando"/>
        </div>
      </div>    
    </div>
    <div id="myModal" class="modal">
      <div class="modal-content">
        <p>¿Estás seguro(a) de eliminar tu publicación?</p>
        <button id="yes">Sí</button>
        <button id="not">No</button>
      </div>
    </div>
    `;

  const mainDashboard = document.createElement('div');
  mainDashboard.classList.add('main-dashboard');
  mainDashboard.innerHTML = viewDashboard;
  const userName = mainDashboard.querySelector('#userName');
  const buttonOut = mainDashboard.querySelector('#button-signOut');
  const postText = mainDashboard.querySelector('#post-text');
  const msgPosT = mainDashboard.querySelector('#msg-post');
  const btnPost = mainDashboard.querySelector('#button-post');
  const containerPost = mainDashboard.querySelector('#containerPosts');
  const myModal = mainDashboard.querySelector('#myModal');
  const not = mainDashboard.querySelector('#not');
  const yes = mainDashboard.querySelector('#yes');
  not.addEventListener('click', () => {
    myModal.style.display = 'none';
  });
  let emailCutted;

  const cutted = (user) => {
    const emailName = user.email;
    const emailAt = emailName.search('@');
    emailCutted = emailName.slice(0, emailAt);
    userName.innerHTML = emailCutted;
    return emailCutted;
  };

  onAuthStateChanged(auth, cutted);

  // funcion para traer el usuario
  const userAuthor = (idAuthor) => {
    if (emailCutted === idAuthor) {
      return true;
    }
    return false;
  };

  let editStatus = false;
  let id = '';
  // savePost();
  gettingPosts((posts) => {
    const postTemplates = posts.map((post) => {
      const dataPost = post.data();
      console.log(dataPost);
      const taskContainerPost = `
        <div class="box-gradient">
          <div id="postPublic">
            <span id="name-post" class="name-user-post">${dataPost.idUser}</span>
              <p id="description-post">
              ${dataPost.post}
              </p>
            <button class="btn-edit icon-pencil" data-id="${post.id}" data-user ="${dataPost.idUser}">
              <i class="fa-solid fa-pencil"></i>
            </button>
            <button class="btn-delete icon-trash" data-id="${post.id}" data-user ="${dataPost.idUser}">
              <i class="fa-solid fa-trash-can"></i>
            </button>
            <button class="btn-like icon-star" data-id="${post.id}">
              <span class="counterLikes" data-id="${post.id}">${dataPost.likes.length}</span>
              <i class="fa-solid fa-star animated" data-id="${post.id}"></i>
            </button>
          </div>
        </div> 
        `;

      return taskContainerPost;
    });

    containerPost.innerHTML = postTemplates.join('');

    // array de strings
    const btnsDelete = containerPost.querySelectorAll('.btn-delete');
    btnsDelete.forEach((btn) => {
      btn.addEventListener('click', () => {
        const idPost2 = btn.dataset.id;
        myModal.style.display = 'block';
        yes.addEventListener('click', () => {
          deletePost(idPost2);
          myModal.style.display = 'none';
        });
      });
      const btnDeleteUser = btn.dataset.user;
      if (userAuthor(btnDeleteUser)) {
        btn.style.display = 'block';
      }
    });
    const btnsEdit = containerPost.querySelectorAll('.btn-edit');
    btnsEdit.forEach((btn) => {
      btn.addEventListener('click', async () => {
        postText.focus();
        const idPost = btn.dataset.id;
        const post = await getPost(idPost);
        const dataPost = post.data();
        postText.value = dataPost.post;
        editStatus = true;
        id = idPost;
        btnPost.innerHTML = 'Guardar';
      });
      const btnEditUser = btn.dataset.user;
      if (userAuthor(btnEditUser)) {
        btn.style.display = 'block';
      }
    });

    const btnsLikes = containerPost.querySelectorAll('.btn-like');
    btnsLikes.forEach((btn) => {
      btn.addEventListener('click', (event) => {
        onAuthStateChanged(auth, async (user) => {
          const idPost3 = event.target.dataset.id;
          const post = await getPost(idPost3);
          const dataPost = post.data();
          const likes = dataPost.likes;
          if (likes.includes(user.email)) {
            await dislikePost(idPost3, user.email);
          } else {
            await likePost(idPost3, user.email);
          }
        });
      });
    });
  });
  btnPost.addEventListener('click', () => {
    if (!editStatus) {
      savePostFire(postText);
      msgPosT.style.display = 'block';
      setTimeout(() => {
        msgPosT.style.display = 'none';
      }, 2000);
    } else {
      editPost(id, { post: postText.value });
      editStatus = false;
      btnPost.innerHTML = 'Publicar';
      postText.value = '';
      msgPosT.style.display = 'block';
      msgPosT.innerHTML = 'Se ha editado tu publicación';
      setTimeout(() => {
        msgPosT.style.display = 'none';
      }, 2000);
    }
  });
  buttonSignOut(buttonOut, navigateTo);
  return mainDashboard;
};
export default Dashboard;
