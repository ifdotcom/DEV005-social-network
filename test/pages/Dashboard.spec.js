/**
 * @jest-environment jsdom
 */
import * as FIREBASE from 'firebase/auth';
import Dashboard from '../../src/pages/Dashboard.js';
// import * as emailCutter from '../../src/pages/Dashboard.js';
import * as savePostF from '../../src/lib/Posts.js';
import { deletePost } from '../../src/lib/firebase.js';

jest.mock('../../src/lib/firebase.js');
jest.mock('firebase/auth');
jest.mock('../../src/images.js', () => ({
  cow: '',
  catGitHub: '',
  logoAlien: '',
  alienMoon: '',
  google: '',
  cowTwo: '',
  imgAlien: '',
  errorImg: '',
}));
describe('Dashboard', () => {
  it('Debería ser una función', () => {
    expect(typeof Dashboard).toBe('function');
  });
  it('Have a buttom', () => {
    FIREBASE.onAuthStateChanged.mockImplementation(() => {});
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    const haveAbutton = DOM.querySelector('#button-signOut');
    expect(haveAbutton).not.toBe(undefined);
  });
  test('After click button signOut return call function navigateTo', (done) => {
    FIREBASE.signOut.mockImplementation(() => Promise.resolve());
    const DOM = document.createElement('div');
    const navigateTo = jest.fn();
    DOM.append(Dashboard(navigateTo));
    const buttonSignOut = DOM.querySelector('#button-signOut');
    buttonSignOut.click();
    // expect(navigateTo).toHaveBeenCalledTimes(1);
    setTimeout(() => {
      expect(navigateTo).toHaveBeenLastCalledWith('/');
      done();
    }, 0);
  });
  test('snapshot of Dashboard', () => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    expect(DOM).toMatchSnapshot();
  });
  test.skip('After click "NO" Modal display change to none', () => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    const not = DOM.querySelector('#not');
    const myModal = DOM.querySelector('#myModal');
    not.click();
    expect(myModal.style.display).toBe('none');
  });
  it.skip('botón de borrar, debería eliminar post con click', (done) => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    // jest.spyOn(savePostF, 'savePostFire');
    const btnsDelete = DOM.querySelectorAll('.btn-delete');
    const myModal = DOM.querySelector('#myModal');
    const yes = DOM.querySelector('#yes');
    btnsDelete[0].click();
    expect(myModal.style.display).toBe('block');
    yes.click();
    expect(deletePost).toHaveBeenCalledTimes(1);
    done();
  });
  it('botón de postear, debería guardar post con click', (done) => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    jest.spyOn(savePostF, 'savePostFire');
    const btnPost = DOM.querySelector('#button-post');
    const postText = DOM.querySelector('#post-text');
    btnPost.click();
    expect(savePostF.savePostFire).toHaveBeenCalledWith(postText);
    done();
  });
  it('boton de postear debería avisar que se posteó con éxito', () => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    const btnPost = DOM.querySelector('#button-post');
    const msgPosT = DOM.querySelector('#msg-post');
    msgPosT.style.display = 'block';
    jest.useFakeTimers();
    btnPost.click();
    jest.advanceTimersByTime(2000);
    expect(msgPosT.style.display).toBe('none');
    jest.useRealTimers();
  });
  it('botón deberia editar al botón, textArea y mensaje ', () => {
    const DOM = document.createElement('div');
    DOM.append(Dashboard());
    const btnPost = DOM.querySelector('#button-post');
    const postText = DOM.querySelector('#post-text');
    const msgPosT = DOM.querySelector('#msg-post');
    jest.useFakeTimers();
    btnPost.click();
    expect(btnPost.innerHTML).toBe('Publicar');
    expect(postText.value).toBe('');
    expect(msgPosT.style.display).toBe('block');
    jest.advanceTimersByTime(3000);
    expect(msgPosT.style.display).toBe('none');
    jest.useRealTimers();
  });
});
