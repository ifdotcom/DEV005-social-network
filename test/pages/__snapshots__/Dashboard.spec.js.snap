// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard snapshot of Dashboard 1`] = `
<div>
  <div
    class="main-dashboard"
  >
    
    
    <div
      id="headerDashboard"
    >
      
      
      <h1
        class="tittle-dashboard"
      >
        Frikis Del Espacio
      </h1>
      
      
      <article
        id="LogOut"
      >
        
        
        <img
          alt="Imagen  "
          id="logoSignOut"
          src=""
        />
        
        
        <button
          id="button-signOut"
        >
          Cerrar Sesión
        </button>
        
      
      </article>
      
    
    </div>
    
    
    <div
      id="barDashboard"
    />
    
    
    
    <div
      id="profile"
    >
      
    
      <div
        id="circleAlien"
      >
        
      
        <div
          id="circleAlien2"
        >
          
        
          <img
            alt="Imagen de Alien "
            id="imgAlien"
            src=""
          />
          
      
        </div>
        
    
      </div>
      
    
      <span
        id="userName"
      >
         Gabriela79 
      </span>
      
    
    </div>
    
    
    <div
      id="border-textArea"
    >
      
      
      <textarea
        id="post-text"
        name="textarea"
        placeholder="Escribir publicación..."
      />
      
    
    </div>
    
    
    <div
      id="msg-post"
    >
      
      
      <span
        id="msg-post-text"
      >
        Se ha compartido tu publicación
      </span>
      
    
    </div>
    
    
    <div
      id="container-btn"
    >
      
      
      <button
        class="btn-post"
        id="button-post"
      >
        Publicar
      </button>
      
    
    </div>
    
    
    <div
      id="containerPosts"
    >
      
      
      <div
        class="box-gradient"
      >
        
        
        <div
          id="postPublic"
        >
          
          
          <span
            class="loading"
          >
            Cargando...
          </span>
          
          
          <img
            alt="Ovni girando"
            id="ovniL"
            src=""
          />
          
        
        </div>
        
      
      </div>
          
    
    </div>
    
    
    <div
      class="modal"
      id="myModal"
    >
      
      
      <div
        class="modal-content"
      >
        
        
        <p>
          ¿Estás seguro(a) de eliminar tu publicación?
        </p>
        
        
        <button
          id="yes"
        >
          Sí
        </button>
        
        
        <button
          id="not"
        >
          No
        </button>
        
      
      </div>
      
    
    </div>
    
    
  </div>
</div>
`;
